keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1537 @@
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
+
+For a resource plugin the following components are needed:
+- Some utility library similar to
+  https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner
+  and the code in driver-registrar.
+- *Resource controller*: a central component which handles resource allocation
+  by watching and modifying ResourceClaims.
+- *Resource node plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by plugins, therefore it is not described further in this KEP.
+
+
+```
+<<[UNRESOLVED @pohly]>>
+All of the changes in Kubernetes need to be specified in a lot more detail.
+
+Upgrade and downgrade scenarios should already be considered for v1alpha1 to ensure
+that whatever changes will be needed are in place before going to v1beta1 where
+downgrades have to be supported.
+
+All of that will be added once there is consensus to move ahead with this proposal.
+<<[/UNRESOLVED]>>
+```
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
+The flow is similar for a ResourceClaim that gets created as a stand-alone
+object by the user. In that case, the Pod reference that ResourceClaim by
+name. The ResourceClaim does not get deleted at the end and can be reused by
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the plugin).
+
+### API
+
+ResourceClaim, ResourceClass and ResourcePlugin are new built-in types
+in a new `cdi.k8s.io/v1alpha1` API group. This was chosen instead of
+using CRDs because core Kubernetes components must interact with them
+and installation of CRDs as part of cluster creation is an unsolved
+problem.
+
+The PodSpec gets extended.
+
+```
+type ResourceClass struct {
+    // Resource plugins have a unique name in reverse domain order (acme.example.com).
+    PluginName string
+    // Parameters holds arbitrary values that will be available to the plugin
+    // when allocating a resource that uses this class. The plugin will
+    // be able to distinguish between parameters stored here and
+    // and those stored in ResourceClaimSpec. These parameters
+    // here can only be set by cluster administrators.
+    Parameters runtime.RawExtension
```

Comment:
The PluginName already points to a plugin which has field "Parameters". Why do we have a "Parameters" field here again?

Reply:
The "PluginName" (now "DriverName") points to a "ResourceDriver" instance. That instance doesn't have parameters.

If you were wondering why both "ResourceClass" and "ResourceClaim" have parameters: that's to support privileged parameters that only an admin is allowed to set and "normal" parameters that users may set.

Reply:
It will be good to add the "privileged parameters" and "normal parameters" clarification to the KEP

Reply:
Isn't "These parameters here can only be set by cluster administrators." already covering this?

Reply:
I've tried to clarify this better when introducing ResourceClass parameters in the summary.